### DISCUSSION TITLE
Encoder script upload to Teensy

### DISCUSSION THREAD

**Discussion by questrov** [fjp/2wd-robot#48] [Q&A] [answered]
Having issues getting the encoder.ino and/or cpp file uploaded to Teensy though managed to get encoder_test running OK.  Can't quite work out how to add the required include files in either Arduino IDE or PlatformIO.  New to PIO and haven't spent the time getting over the learning curve yet...

Want to get the ROS nodes going, SO CLOSE!  Can someone please give me a clue?

**Comment by fjp** [comment]
Hi @questrov, have you tried to follow the instructions for Arduino IDE outline here in the documentation: https://ros-mobile-robots.com/teensy-mcu/ and what is the error you are getting?

The [`encoders.ino`](https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base/scripts/encoders/encoders/encoders.ino) was developed for the first version of DiffBot and not for Remo, although it can be used for Remo too. In case you want to use `encoders.ino`, I suggest using the high level PID controllers that are running in the `diffbot_hw_interface`:

https://github.com/ros-mobile-robots/diffbot/blob/ff78f49bf75c8c5b9f392a929fd9d38fcadb4514/diffbot_base/src/diffbot_hw_interface.cpp#L188-L189

Alternatively you can use the low level PID controllers running on the mcu. Then you don't need the code inside [`encoders`](https://github.com/ros-mobile-robots/diffbot/tree/noetic-devel/diffbot_base/scripts/encoders/encoders) folder and should instead upload the [`base_controller`](https://github.com/ros-mobile-robots/diffbot/tree/noetic-devel/diffbot_base/scripts/base_controller) using PlatformIO.

**Reply by questrov** [reply]
Thanks fjp for the quick reply.  Yes, followed instructions in the link OK.  Got blink and the encoder test to upload OK though can't recall if it was the IDE or PlatformIO that eventually worked for encoder test.  Was late night and flailing at that point:-/

With IDE issue is `fatal error: diffbot_msgs/Encoders.h: No such file or directory
compilation terminated.`  Basically it can't find Encoders.h in the diffbot_msgs folder even though both are in a library folder.  Interesting if I change line from `#include <diffbot_msgs/Encoders.h>` to `#include <Encoders.h>` it finds the Encoders.h file if I put it in the library folder by itself.  Then of course the ROS topics don't match.
![image](https://user-images.githubusercontent.com/39848760/149597481-e90c5acb-5052-4512-a553-cc425ed89cb6.png)

Not sure what I'm doing wrong.  Really need to work out how to upload with PlatformIO but it's not exactly intuitive, at least to me.  I always struggle the first time until I get it with new (to me) software.

I'm using the Adafruit Feather motor controller wired up per the Remo schematic, does that mean I should be using the base_controller anyway?

**Comment by fjp** [comment, marked as answer]
> With IDE issue is fatal error: diffbot_msgs/Encoders.h: No such file or directory compilation terminated. Basically it can't find Encoders.h in the diffbot_msgs folder even though both are in a library folder. Interesting if I change line from #include <diffbot_msgs/Encoders.h> to #include <Encoders.h> it finds the Encoders.h file if I put it in the library folder by itself. Then of course the ROS topics don't match.

Have you built the custom `diffbot_msgs` for arduino rosserial? The procedure how to do this should be explained (hopefully well enough) here https://ros-mobile-robots.com/diffbot_msgs/#rosserial. If it doesn't help please let me know. 

> I'm using the Adafruit Feather motor controller wired up per the Remo schematic, does that mean I should be using the base_controller anyway?

Then you most likely don't need `encoder.ino` and I suggest you go with the code in `base_controller` for the mcu. The docs on this approach are basically non existent at the moment... You will need VSCode with PlatformIO installed (I guess you have that already). Then it is important to open the `base_controller` folder in VSCode (File -> Open Folder ...). VSCode/PlatformIO should then be able to recongize the `pltaformio.ini` file inside `base_controller` folder. I just uploaded a video showing the process here in the documentation: https://ros-mobile-robots.com/teensy-mcu/#base-controller.

**Reply by questrov** [reply]
Thanks for the clues, much appreciated!  Negative on custom diffbot_msgs for arduino rosserial.  I've basically followed the instructions on the diffbot GitHub page to the point of getting Gazebo simulation with ROS control working.  Also have A1 Lidar, BNO055 and USB cam working independently in ROS on Pi4.  Motors with quad encoders finally arrived and tested that's about it.  Next step is getting it all working at the same time on the bench:-/

Will watch the video and take a bit more time on the PlatformIO learning curve.  Thanks again for the help will let you know how it goes...

**Reply by questrov** [reply]
`code .` that's the clue I was missing, many thanks for the video!  Upload OK got wheel ticks and bringup minimal and/or laser launch OK, yay!

Now that the encoder question is officially answered I have about 100 more:-/  Need to read the docs again to be sure the answer is not there before asking...

Next hurdle is the Adafruit FeatherWing...

**Reply by fjp** [reply]
Glad to hear you got a step further. You are also welcome to ask, the docs are not fully complete, especially for Remo and the Adafruit FeatherWing. You can just create a new question if you need help.

And by the way, you might find what you are looking for using the search feature at the top of the [docs](https://ros-mobile-robots.com/) - but I guess nothing useful for Adafruit FeatherWing yet.
